File: web-go/internal/service/auth_service.go
package service

import (
	"context"
	"strings"

	"agussyahrilmubarok.github.io/web/internal/domain"
	"agussyahrilmubarok.github.io/web/internal/model"
	"agussyahrilmubarok.github.io/web/internal/repository"
	"agussyahrilmubarok.github.io/web/pkg/cryptoutil"
	"agussyahrilmubarok.github.io/web/pkg/logger"
	"github.com/google/uuid"
	"go.uber.org/zap"
)

//go:generate mockery --name=IAuthService
type IAuthService interface {
	SignUp(ctx context.Context, param model.SignUpRequest) (*model.UserResponse, error)
	SignIn(ctx context.Context, param model.SignInRequest) (*model.UserResponse, error)
}

type authService struct {
	userRepository repository.IUserRepository
}

// SignUp implements [IAuthService].
func (s *authService) SignUp(ctx context.Context, param model.SignUpRequest) (*model.UserResponse, error) {
	log := logger.FromCtx(ctx).With(zap.String("user_email", param.Email))

	exist, err := s.userRepository.ExistsByEmailIgnorecase(ctx, param.Email)
	if err != nil {
		log.Error("failed to check existing user email", zap.Error(err))
		return nil, err
	}

	if exist {
		log.Warn("user email already exists")
		return nil, domain.ErrUserEmailExists
	}

	hashPassword, err := cryptoutil.PasswordHash(param.Password)
	if err != nil {
		log.Error("failed to hash user password", zap.Error(err))
		return nil, err
	}

	user := &domain.User{
		ID:       uuid.New(),
		Name:     param.Name,
		Email:    strings.ToLower(param.Email),
		Password: hashPassword,
	}

	if err := s.userRepository.Create(ctx, user); err != nil {
		log.Error("failed to create user", zap.Error(err))
		return nil, err
	}

	log.Info("user signup completed successfully", zap.String("user_id", user.ID.String()))
	userResponse := model.ToUserResponse(user)
	return &userResponse, nil
}

// SignIn implements [IAuthService].
func (s *authService) SignIn(ctx context.Context, param model.SignInRequest) (*model.UserResponse, error) {
	log := logger.FromCtx(ctx).With(zap.String("user_email", param.Email))

	user, err := s.userRepository.FindByEmail(ctx, strings.ToLower(param.Email))
	if err != nil {
		log.Error("failed to find user by email", zap.Error(err))
		return nil, err
	}
	if user == nil {
		log.Warn("user email not found")
		return nil, domain.ErrUserEmailNotFound
	}

	if err := cryptoutil.PasswordCheck(param.Password, user.Password); err != nil {
		log.Warn("invalid user password", zap.String("user_id", user.ID.String()))
		return nil, domain.ErrUserPasswordNotMatch
	}

	log.Info("user signin completed successfully", zap.String("user_id", user.ID.String()))
	userResponse := model.ToUserResponse(user)
	return &userResponse, nil
}

func NewAuthService(
	userRepository repository.IUserRepository,
) IAuthService {
	return &authService{
		userRepository: userRepository,
	}
}

File: examples/web-mvc-i18n/web-dotnet/Web/Application/Service/AuthService.cs
using Web.Application.DTO.Auth;
using Web.Application.DTO.User;
using Web.Common.Exceptions;
using Web.Domain.User;

namespace Web.Application.Service;

public class AuthService : IAuthService
{

    private readonly IUserRepository _userRepository;
    private readonly ILogger<AuthService> _logger;

    public AuthService(IUserRepository userRepository, ILogger<AuthService> logger)
    {
        _userRepository = userRepository;
        _logger = logger;
    }

    public async Task<UserResponse> SignUpAsync(SignUpRequest request, CancellationToken ct = default)
    {
        if (await _userRepository.ExistsByEmailAsync(request.Email.ToLowerInvariant(), ct))
        {
            _logger.LogWarning("Email already taken: {Email}", request.Email);
            throw new EmailAlreadyExistsException();
        }

        var hashedPassword = BCrypt.Net.BCrypt.HashPassword(request.Password);
        var user = User.Create(request.Name, request.Email.ToLowerInvariant(), hashedPassword);
        await _userRepository.CreateAsync(user, ct);

        _logger.LogInformation("Signed up successfully {Id}", user.Id);
        return ToResponse(user);
    }

    public async Task<UserResponse> SignInAsync(SignInRequest request, CancellationToken ct = default)
    {
        var user = await _userRepository.FindByEmailAsync(request.Email.ToLowerInvariant(), ct);

        if (user is null || !BCrypt.Net.BCrypt.Verify(request.Password, user.Password))
        {
            _logger.LogWarning("Sign in failed for email: {Email}", request.Email);
            throw new InvalidCredentialsException();
        }

        _logger.LogInformation("Signed in successfully {Id}", user.Id);
        return ToResponse(user);
    }

    private static UserResponse ToResponse(User user) =>
        new(user.Id, user.Name, user.Email, user.CreatedAt, user.UpdatedAt);
}